src/generate.ts

```diff
@@ -0,0 +1,192 @@
+		}
+	}
+
+	switch (f.dataType) {
+		case 'ConceptType':
+		case 'ForeignKey': {
+			const referencedInterface = getReferencedInterface(
+				m.tables[f.references!.resourceName].name,
+				mode,
+			);
+			const referencedFieldType = `${referencedInterface}['${f.references!.fieldName}']`;
+			if (mode === 'Write') {
+				return referencedFieldType;
+			}
+
+			const nullable = f.required ? '' : '?';
+			return `{ __id: ${referencedFieldType} } | [${referencedInterface}${nullable}]`;
+		}
+		default:
+			return `Types['${f.dataType}']['${mode}']`;
+	}
+};
+
+const fieldsToInterfaceProps = (
+	m: RequiredModelSubset,
+	fields: AbstractSqlField[],
+	mode: Mode,
+): string[] =>
+	fields.map((f) => {
+		const nullable = f.required ? '' : ' | null';
+		return `${sqlNameToODataName(f.fieldName)}: ${sqlTypeToTypescriptType(
+			m,
+			f,
+			mode,
+		)}${nullable};`;
+	});
+
+const recurseRelationships = (
+	m: RequiredModelSubset,
+	relationships: Relationship,
+	inverseSynonyms: Record<string, string>,
+	mode: Mode,
+	currentTable: AbstractSqlTable,
+	parentKey: string,
+): string[] =>
+	Object.keys(relationships).flatMap((key) => {
+		if (key === '$') {
+			const [localField, referencedField] = (
+				relationships as RelationshipLeafNode
+			).$;
+			if (currentTable.idField === localField && referencedField != null) {
+				const referencedTable = m.tables[referencedField[0]];
+				if (referencedTable != null) {
+					const referencedInterface = getReferencedInterface(
+						referencedTable.name,
+						mode,
+					);
+					const propDefinitons = [
+						`${parentKey}?: Array<${referencedInterface}>;`,
+					];
+					const synonym = inverseSynonyms[odataNameToSqlName(parentKey)];
+					if (synonym != null) {
+						propDefinitons.push(
+							`${sqlNameToODataName(synonym)}?: Array<${referencedInterface}>;`,
+						);
+					}
+					return propDefinitons;
+				}
+			}
+			return [];
+		}
+		return recurseRelationships(
+			m,
+			(relationships as RelationshipInternalNode)[key],
+			inverseSynonyms,
+			mode,
+			currentTable,
+			`${parentKey}__${key.replace(/ /g, '_')}`,
+		);
+	});
+
+const relationshipsToInterfaceProps = (
+	m: RequiredModelSubset,
+	table: AbstractSqlTable,
+	mode: Mode,
+): string[] => {
+	const relationships = m.relationships[table.resourceName];
+	if (relationships == null) {
+		return [];
+	}
+	return Object.keys(relationships).flatMap((key) => {
+		// We skip `has` a the top level as we omit it by convention
+		if (key === 'has') {
+			return [];
+		}
+		const inverseSynonyms = Object.fromEntries(
+			Object.entries(m.synonyms).map(([termForm, factType]) => [
+				factType,
+				termForm,
+			]),
+		);
+		return recurseRelationships(
+			m,
+			relationships[key],
+			inverseSynonyms,
+			mode,
+			table,
+			key.replace(/ /g, '_'),
+		);
+	});
+};
+
+const tableToInterface = (m: RequiredModelSubset, table: AbstractSqlTable) => {
+	return trimNL`
+export interface ${modelNameToCamelCaseName(table.name)} {
+	Read: {
```

Comment:
Not-a-deal-breaker: I'm wondering whether the result would be more readable if we were emitting a Read & a Write namespace rather than using interface props.
That way we could avoid the repetition of Read/Write on each prop (eg `created_at: Types['Date Time']['Read'];`).
I realize though that
* after referencing TestTag['Read'], then from that point onward we no longer need to bother w/ Read/Write any more.
* it might be a worse UX when we need to import & use both the Read & Write variants of a resource in a file b/c of conflicts.

Reply:
So the key property I need is that you can have things like
```ts
const fn = <T extends Resource>(x: T['Read'], y: T['Write']) => ...
```
where you can pass in a single type to a generic and access both read/write typings as necessary. If there is a way that allows that whilst making the typings a bit more concise then I'm happy to do so but I don't fully understand your suggestion/how it would work whilst keeping that property